src/c-writer.cc

```diff
@@ -527,87 +546,205 @@ std::string CWriter::MangleTagTypes(const TypeVector& types) {
-      }
-    }
-  }
-
-  return result;
+/* The C symbol for an export from this module. */
+std::string CWriter::ExportName(std::string_view export_name) {
+  return kSymbolPrefix + module_prefix_ + '_' + MangleName(export_name);
 }
 
-std::string CWriter::ExportName(std::string_view mangled_name) {
-  return module_prefix_ + std::string(mangled_name);
+/* The C symbol for an export from an arbitrary module. */
+// static
+std::string CWriter::ExportName(std::string_view module_name,
+                                std::string_view export_name) {
+  return kSymbolPrefix + MangleModuleName(module_name) + '_' +
+         MangleName(export_name);
 }
 
+/* The type name of an instance of this module. */
 std::string CWriter::ModuleInstanceTypeName() const {
-  return module_prefix_ + "_instance_t";
+  return kSymbolPrefix + module_prefix_;
 }
 
+/* The type name of an instance of an arbitrary module. */
 // static
-std::string CWriter::MangleModuleInstanceName(std::string_view module_name) {
-  return MangleName(module_name) + "_instance";
+std::string CWriter::ModuleInstanceTypeName(std::string_view module_name) {
+  return kSymbolPrefix + MangleModuleName(module_name);
 }
 
 // static
-std::string CWriter::MangleModuleInstanceTypeName(
-    std::string_view module_name) {
-  return MangleName(module_name) + "_instance_t";
-}
+std::string CWriter::Mangle(std::string_view name, bool double_underscores) {
+  /*
+   * Name mangling transforms arbitrary Wasm names into "safe" C names
+   * in a deterministic way. To avoid collisions, distinct Wasm names must be
+   * transformed into distinct C names.
+   *
+   * The rules implemented here are:
+   * 1) any hex digit ('A' through 'F') that follows the sequence "0x"
+   *    is escaped
+   * 2) any underscore at the beginning, at the end, or following another
+   *    underscore, is escaped
+   * 3) if double_underscores is set, underscores are replaced with
+   *    two underscores.
+   * 4) otherwise, any alphanumeric character is kept as-is,
+   *    and any other character is escaped
+   *
+   * "Escaped" means the character is represented with the sequence "0xAB",
+   * where A B are hex digits ('0'-'9' or 'A'-'F') representing the character's
+   * numeric value.
+   *
+   * Module names are mangled with double_underscores=true to prevent
+   * collisions between, e.g., a module "alfa" with export
+   * "bravo_charlie" vs. a module "alfa_bravo" with export "charlie".
+   *
+   * In the implementation, we don't use the standard isalpha/isdigit/isalnum
+   * because the caller might have changed the current locale.
+   */
+
+  auto my_isalpha = [](uint8_t ch) {
+    return (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z');
+  };
+  auto my_isdigit = [](uint8_t ch) { return (ch >= '0' && ch <= '9'); };
+  auto my_isalnum = [&](uint8_t ch) {
+    return my_isalpha(ch) || my_isdigit(ch);
+  };
+  auto my_ishexdigit = [&](uint8_t ch) {
+    return my_isdigit(ch) || (ch >= 'A' && ch <= 'F');
+  };
```

Comment:
How about simply declaring these as normal static C functions outside this method body?

I guess its mostly aesthetic, but unless we actually need to capture stuff from the current function I'm not a huge fan of inner lambda functions like this.

Reply:
:+1:

Reply:
Done in 6974baef56330fb2d550aa2ede8a7f4fbfe71d0d

Reply:
Thank you!